File: MIDTERM-LAB-3/mp03_search_keyword.js
/**
 * MP03 - Search for a Keyword in the Dataset
 * Dataset: Pearson VUE Exam Results - University of Perpetual Help System Molino
 * Columns: Candidate, Student/Faculty/NTE, Exam, Language, Exam Date, Score, Result, Time Used
 *
 * Program Logic:
 * Prompts for a CSV path and search keyword. Skips the Pearson VUE preamble,
 * then performs a case-insensitive search across every meaningful field of every
 * data row. Matching rows are displayed in a formatted table with a "Matched In"
 * column. Long values are truncated to keep the table aligned on a standard terminal.
 */

const fs       = require('fs');
const readline = require('readline');

// ── Column display configuration ─────────────────────────────────────────────
const COL_CONFIG = [
    { key: 'Candidate',              maxW: 20 },
    { key: 'Student/ Faculty/ NTE',  maxW: 10 },
    { key: 'Exam',                   maxW: 30 },
    { key: 'Language',               maxW: 8  },
    { key: 'Exam Date',              maxW: 10 },
    { key: 'Score',                  maxW: 5  },
    { key: 'Result',                 maxW: 6  },
    { key: 'Time Used',              maxW: 13 },
];
const MATCH_COL_MAXW = 25; // Max width for the "Matched In" column

function parseCSVLine(line) {
    const fields = [];
    let current = '', insideQuote = false;
    for (const char of line) {
        if (char === '"') { insideQuote = !insideQuote; }
        else if (char === ',' && !insideQuote) { fields.push(current.trim()); current = ''; }
        else { current += char; }
    }
    fields.push(current.trim());
    return fields;
}

/**
 * truncate - Clips a string to maxLen, appending … if needed.
 */
function truncate(str, maxLen) {
    const s = String(str);
    return s.length > maxLen ? s.slice(0, maxLen - 1) + '\u2026' : s;
}

function padRight(str, len) { return String(str).padEnd(len, ' '); }

function searchKeyword(filePath, keyword) {
    if (!keyword) { console.error('Error: Keyword cannot be empty.'); process.exit(1); }
    if (!fs.existsSync(filePath)) { console.error(`Error: File not found -> "${filePath}"`); process.exit(1); }

    let raw;
    try { raw = fs.readFileSync(filePath, 'utf8'); }
    catch (err) { console.error(`Error reading file: ${err.message}`); process.exit(1); }

    const lines    = raw.replace(/\uFEFF/g, '').replace(/\r/g, '').split('\n');
    const kwLower  = keyword.toLowerCase();

    let header    = null;
    let colIdxMap = {};   // display key -> CSV column index
    let usedCols  = [];   // all CSV column indices with non-empty header labels

    const allRows     = [];
    const matchedRows = [];
    const matchInfo   = [];

    for (const line of lines) {
        if (line.trim() === '') continue;
        const fields = parseCSVLine(line);

        if (header === null) {
            if (fields[0].toLowerCase() === 'candidate') {
                header = fields;
                // Map display columns to CSV indices
                COL_CONFIG.forEach(col => {
                    const idx = header.findIndex(h => h.trim().toLowerCase() === col.key.toLowerCase());
                    colIdxMap[col.key] = idx;
                });
                // All non-empty header columns (for search across ALL fields)
                usedCols = header.map((h, i) => ({ h, i }))
                                 .filter(({ h }) => h.trim() !== '')
                                 .map(({ i }) => i);
            }
            continue;
        }

        if (fields[0].trim() === '') continue;
        allRows.push(fields);

        // Search every non-empty column for the keyword
        const hits = [];
        usedCols.forEach(colIdx => {
            const val = fields[colIdx] !== undefined ? fields[colIdx] : '';
            if (val.toLowerCase().includes(kwLower)) hits.push(header[colIdx]);
        });

        if (hits.length > 0) {
            matchedRows.push(fields);
            matchInfo.push(truncate(hits.join(', '), MATCH_COL_MAXW));
        }
    }

    if (!header) { console.error('Error: Header row not found.'); process.exit(1); }

    // ── Summary ──────────────────────────────────────────────────────────
    console.log('\n============================================================');
    console.log('     MP03 - Keyword Search Results');
    console.log('     Pearson VUE Exam Results');
    console.log('============================================================');
    console.log(`File     : ${filePath}`);
    console.log(`Keyword  : "${keyword}"`);
    console.log(`Searched : ${allRows.length} records`);
    console.log(`Matches  : ${matchedRows.length} records`);

    if (matchedRows.length === 0) {
        console.log(`\nNo records matched the keyword "${keyword}".`); return;
    }

    // ── Build display rows with truncation ───────────────────────────────
    const colW = COL_CONFIG.map(col => col.key.length);
    let matchColW = 'Matched In'.length;

    const displayRows = matchedRows.map((row, r) => {
        const cells = COL_CONFIG.map((col, i) => {
            const idx = colIdxMap[col.key];
            const raw = (idx >= 0 && idx < row.length) ? row[idx] : '';
            const val = truncate(raw, col.maxW);
            colW[i] = Math.max(colW[i], val.length);
            return val;
        });
        matchColW = Math.max(matchColW, matchInfo[r].length);
        return cells;
    });

    // ── Build separator ──────────────────────────────────────────────────
    const separator = '+' + colW.map(w => '-'.repeat(w + 2)).join('+') + '+' + '-'.repeat(matchColW + 2) + '+';

    // ── Print table ──────────────────────────────────────────────────────
    console.log('\n' + separator);
    console.log('| ' + COL_CONFIG.map((col, i) => padRight(col.key, colW[i])).join(' | ')
        + ' | ' + padRight('Matched In', matchColW) + ' |');
    console.log(separator);

    displayRows.forEach((row, r) => {
        console.log('| ' + row.map((v, i) => padRight(v, colW[i])).join(' | ')
            + ' | ' + padRight(matchInfo[r], matchColW) + ' |');
    });

    console.log(separator);
    console.log(`Total matching rows: ${matchedRows.length}`);
}

// ─── Entry Point ──────────────────────────────────────────────────────────────
const rl = readline.createInterface({ input: process.stdin, output: process.stdout });
console.log('============================================================');
console.log('     MP03 - Search for a Keyword');
console.log('     Pearson VUE Exam Results');
console.log('============================================================');
rl.question('Enter the CSV dataset file path: ', path => {
    rl.question('Enter keyword to search: ', kw => {
        rl.close();
        searchKeyword(path.trim(), kw.trim());
    });
});

File: MIDTERM-LAB-3/MP03_SearchKeyword.java
import java.io.*;
import java.util.*;

/**
 * MP03 - Search for a Keyword in the Dataset
 * Dataset: Pearson VUE Exam Results - University of Perpetual Help System Molino
 *
 * Program Logic:
 * Prompts for a CSV path and a search keyword. Skips the Pearson VUE preamble,
 * then performs a case-insensitive search across every meaningful field of every
 * data row. Matching rows are collected and displayed in a formatted table with
 * a "Matched In" column showing which field(s) triggered the match. All long
 * values are truncated to keep the output aligned in a standard terminal.
 */
public class MP03_SearchKeyword {

    // ── Column display configuration ──────────────────────────────────────────
    static final String[] COL_LABELS = {
        "Candidate", "Student/ Faculty/ NTE", "Exam",
        "Language", "Exam Date", "Score", "Result", "Time Used"
    };
    static final int[] COL_MAX = { 20, 10, 30, 8, 10, 5, 6, 13 };
    static final int MATCH_MAX = 25; // max width for "Matched In" column

    static String[] parseCSVLine(String line) {
        List<String> fields = new ArrayList<>();
        StringBuilder cur = new StringBuilder();
        boolean inQ = false;
        for (char c : line.toCharArray()) {
            if (c == '"')              { inQ = !inQ; }
            else if (c == ',' && !inQ) { fields.add(cur.toString().trim()); cur.setLength(0); }
            else                       { cur.append(c); }
        }
        fields.add(cur.toString().trim());
        return fields.toArray(new String[0]);
    }

    /** truncate - Clips str to maxLen; appends ellipsis if clipped. */
    static String truncate(String str, int maxLen) {
        if (str == null) str = "";
        return str.length() > maxLen ? str.substring(0, maxLen - 1) + "\u2026" : str;
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        System.out.println("============================================================");
        System.out.println("     MP03 - Search for a Keyword                            ");
        System.out.println("     Pearson VUE Exam Results                               ");
        System.out.println("============================================================");
        System.out.print("Enter the CSV dataset file path: ");
        String filePath = scanner.nextLine().trim();
        System.out.print("Enter keyword to search: ");
        String keyword = scanner.nextLine().trim();

        if (keyword.isEmpty()) { System.out.println("Error: Keyword cannot be empty."); scanner.close(); return; }

        String kwLower = keyword.toLowerCase();

        String[] csvHeader = null;
        int[]    colIdx    = new int[COL_LABELS.length]; // CSV column indices for display cols
        int[]    allIdx    = null;                        // All non-empty CSV column indices (for search)
        Arrays.fill(colIdx, -1);

        ArrayList<String[]> allRows     = new ArrayList<>();
        ArrayList<String[]> matchedRows = new ArrayList<>();
        ArrayList<String>   matchInfo   = new ArrayList<>();

        try {
            BufferedReader reader = new BufferedReader(
                new InputStreamReader(new FileInputStream(filePath), "UTF-8"));
            String line;

            while ((line = reader.readLine()) != null) {
                line = line.replace("\uFEFF", "").replace("\r", "");
                if (line.trim().isEmpty()) continue;
                String[] fields = parseCSVLine(line);

                if (csvHeader == null) {
                    if (fields.length > 0 && fields[0].equalsIgnoreCase("Candidate")) {
                        csvHeader = fields;
                        // Map display labels to CSV indices
                        for (int d = 0; d < COL_LABELS.length; d++)
                            for (int c = 0; c < csvHeader.length; c++)
                                if (csvHeader[c].trim().equalsIgnoreCase(COL_LABELS[d])) { colIdx[d] = c; break; }
                        // Build list of ALL non-empty header column indices (for search scope)
                        List<Integer> tmp = new ArrayList<>();
                        for (int c = 0; c < csvHeader.length; c++)
                            if (!csvHeader[c].trim().isEmpty()) tmp.add(c);
                        allIdx = tmp.stream().mapToInt(Integer::intValue).toArray();
                    }
                    continue;
                }

                if (fields[0].trim().isEmpty()) continue;
                allRows.add(fields);

                // Search every non-empty column for the keyword
                List<String> hits = new ArrayList<>();
                for (int ci : allIdx) {
                    String val = (ci < fields.length) ? fields[ci] : "";
                    if (val.toLowerCase().contains(kwLower)) hits.add(csvHeader[ci]);
                }

                if (!hits.isEmpty()) {
                    matchedRows.add(fields);
                    matchInfo.add(truncate(String.join(", ", hits), MATCH_MAX));
                }
            }
            reader.close();

            if (csvHeader == null) { System.out.println("Error: Header row not found."); return; }

            // ── Summary ───────────────────────────────────────────────────
            System.out.println("\nFile     : " + filePath);
            System.out.println("Keyword  : \"" + keyword + "\"");
            System.out.println("Searched : " + allRows.size() + " records");
            System.out.println("Matches  : " + matchedRows.size() + " records");

            if (matchedRows.isEmpty()) {
                System.out.println("\nNo records matched the keyword \"" + keyword + "\"."); return;
            }

            // ── Build display cells with truncation ───────────────────────
            int numCols = COL_LABELS.length;
            int[] colW  = new int[numCols];
            for (int d = 0; d < numCols; d++) colW[d] = COL_LABELS[d].length();
            int matchColW = "Matched In".length();

            String[][] displayRows = new String[matchedRows.size()][numCols];
            for (int r = 0; r < matchedRows.size(); r++) {
                String[] row = matchedRows.get(r);
                for (int d = 0; d < numCols; d++) {
                    int ci  = colIdx[d];
                    String raw = (ci >= 0 && ci < row.length) ? row[ci] : "";
                    String val = truncate(raw, COL_MAX[d]);
                    displayRows[r][d] = val;
                    colW[d] = Math.max(colW[d], val.length());
                }
                matchColW = Math.max(matchColW, matchInfo.get(r).length());
            }

            // ── Build separator ───────────────────────────────────────────
            StringBuilder sb = new StringBuilder("+");
            for (int w : colW) sb.append("-".repeat(w + 2)).append("+");
            sb.append("-".repeat(matchColW + 2)).append("+");
            String sep = sb.toString();

            // ── Print table ───────────────────────────────────────────────
            System.out.println("\n" + sep);
            System.out.print("|");
            for (int d = 0; d < numCols; d++) System.out.printf(" %-" + colW[d] + "s |", COL_LABELS[d]);
            System.out.printf(" %-" + matchColW + "s |%n", "Matched In");
            System.out.println(sep);

            for (int r = 0; r < displayRows.length; r++) {
                System.out.print("|");
                for (int d = 0; d < numCols; d++) System.out.printf(" %-" + colW[d] + "s |", displayRows[r][d]);
                System.out.printf(" %-" + matchColW + "s |%n", matchInfo.get(r));
            }

            System.out.println(sep);
            System.out.println("Total matching rows: " + matchedRows.size());

        } catch (FileNotFoundException e) { System.out.println("Error: File not found -> " + filePath);
        } catch (IOException e)           { System.out.println("Error reading file: " + e.getMessage());
        } finally                         { scanner.close(); }
    }
}
